### ISSUE TITLE
Make `crf_baseline_scripts/run.sh` conform to peval specs

### ISSUE THREAD

**Issue by mhuesch** [GaloisInc/ppaml-cp6#15]
In the Submission Requirements document (https://owncloud.galois.com/public.php?service=files&t=ba41c2b2d994cb05f0d3756816f37929) we specify that the `run.sh` file should take 4 arguments (see page 9) which are paths to a parameter file, the input directory, the output directory, and the log directory. The input & output directories correspond to the run_in and run_out directories in the sandboxes. I am having trouble getting the CRF to run with peval because it has hardcoded paths relative to the directory where `run.sh` is called, and also because it requires files from eval_in (which aren't provided to the solutions at execution time because then teams could cheat and read the ground truth).

Chengjiang - I assigned this to you because you last worked on the CRF, but I don't know if Roddy or Eran would be better suited to take care of this.

Could you modify the solution to conform with the submission requirements, and to not expect files from eval_in? The former is likely as simple as passing an argument in `run.sh` to the `matlab ...` invocation. The latter I am not sure.

Making this change would allow us to work with the solution more easily, using peval. It would make data collection and comparisons easier. Thanks.

**Comment by chengjianglong** [comment]
Michael, to meet the requirement, we can merge the directory eval_in under the directory run_in.  I will modify this today.

**Comment by mhuesch** [comment]
@chengjianglong So I am curious then - why does the reference solution need access to files in eval_in/, which are not available to submissions for the performer's solutions? For example CRA's CP6 solution ran fine with only access to run_in/. Merging eval_in/ with run_in/ certainly works, but I wonder if it is the best way (given that it seems possible to manage with only the run_in/ files and indeed round 1 was conducted on that assumption).

If we do merge run_in/ & eval_in/ we should probably update the datasets on Midas and announce the change, which is why I hesitate to do that.

**Comment by chengjianglong** [comment]
@mhuesch I see your points. eval_in/ is designed to evaluate the performance.  In our current implement, we conduct the training and testing parts so that we need the eval_in to evaluate the performance for the testing purpose.  

If we keep the eval_in/, and then that' means only we have the condition to evaluate the performance, given the prediction results.  This is a common rule that public datasets follow to avoid cheating.

However, if we only want to provide the baseline results,  sharing the curve data of the baselines is enough.  If we want to share the code,  we'd better share both the training and testing parts so that the users can evaluate the performances by the eval_in/ that also can be generated by the users.

Hopefully this can help solve your concerns.

**Comment by chengjianglong** [comment]
@mhuesch To clarify, what I mean about "merge"  is to put run_in/ and eval_in/ into the "input_directory" for the run.sh.  We can point out that eval_in/ should be generated by the users if they want to evaluate their developed algorithms using their own testing data on their sides.  

However,  to evaluate on the public testing data,  they should submit their prediction results to us, because only we have the corresponding eval_in/.

**Comment by mhuesch** [comment]
@chengjianglong I am not sure I follow. Does that mean inference checks its results against ground truth? If so, wouldn't that give it an advantage over the other teams' solutions? The other teams can only run inference over the 'testing' data, but cannot tell if the results are "good" or not...

I don't see why the `run.sh` solution needs access to eval_in/ - usually only the evaluator (`eval.sh`) is able to access that. Perhaps I am misunderstanding you though.

For the public dataset, the teams can view the whole thing (thus they have access to eval_in/). But for the private dataset we don't want to reveal eval_in/ to their solution, because then they would be able to see the ground truth. This is why merging run_in/ & eval_in/ is problematic.

**Comment by mhuesch** [comment]
If it would be easier to discuss on the phone or google hangout, I am available tomorrow if that works for you! That might be more efficient than messaging.

**Comment by chengjianglong** [comment]
@mhuesch I see your points.  We can change the implementation to let run.sh conduct both the training and testing parts, without checking against ground truth.  

Actually, we were talking the same things.  The "run.sh" that I thought before is combining the "run.sh" you thought and the "eval.sh" together in a single bash file.  To reach the consistence,  I can separate the "eval.sh" part from the current implement of "run.sh".

Thanks for clarification.

**Comment by mhuesch** [comment]
Ah, ok great! Yes, the `run.sh` script should only do training+testing, but not compare to ground truth.

Does that mean that https://github.com/GaloisInc/ppaml-cp6/blob/master/data_preparation_tools/eval.sh is contained in `run.sh`? That file is the evaluator I used for evaluation round-1 solutions.

**Comment by chengjianglong** [comment]
@mhuesch Yes, we contained the eval.sh part n our current run.sh so that's why we need eval_in.

**Comment by chengjianglong** [comment]
Yes, we contained the eval.sh part in our current run.sh so that's why we need eval_in.

**Comment by collinsr** [comment]
Michael,

The files we need from eval_in are all in the eval_in/etc directory, and are global across all of CP6 (i.e. not specific to any partition of the data for training or testing.) The sandbox_adapter.py script uses these files when converting the sandbox to McAuley's format.

Has the MATLAB CRF ever run in your peval framework? If so, can you send us the run.sh you used?

It sounds like what we need is for the first argument to your script (the parameter file) to contain a path to the files in eval_in/etc , which is all our training step needs. The second argument is the path to run_in, the third is the path to run_out, and the fourth will be the logging directory. The logging directory will probably be used something like

matlab -nodisplay -other-args -r "run('path-to-wrapper.m')" &> $LOGGING_DIR/run.log

or something like that.

If that sounds right, then all we need is the path to eval_in/etc (or another directory containing those files) in the parameter file in the first argument.

Thanks,
-Roddy

**Comment by chengjianglong** [comment]
@mhuesch We also need to make sure the first argument containing the path to the directory of "crf_marbl" to run this MATLAB CRF.

**Comment by mhuesch** [comment]
@collinsr: Got it. I don't believe the CRF has ever run in peval, because it wasn't able to run without eval_in/ (and because it wasn't able to take paths specifying run_in/, run_out/, and log/ as arguments). It seems like we have two options: 1) move the eval_in/etc/ code somewhere globally accessible and put the path in the config file (along with the "crf_marbl" path) or 2) remove the evaluation step from `run.sh`, which would remove the need for eval_in/. Either of these is fine with me - (2) seems neater but doing the simpler thing (whichever that is) seems fine to me.

**Comment by chengjianglong** [comment]
@mhuesch I deal with this by adding a configuration file (See: https://github.com/GaloisInc/ppaml-cp6/blob/dev/crf-changes/crf_baseline_scripts/configure.txt), which can be taken as the first argument.
